##3699938 Epic Comments

andela-oowonikoko opened the pull request:
### What does this PR do ?

- Create comments for a particular contribution
- Fetch all the comments that belong to a contribution
- Implements deletetion of single comment

### Description of Task to be completed ? 
- Create an endpoint for creating comments by a contribution id  
   `/contributions/:contribution_id/comments`
- Create an endpoint for fetching comments by a contribution id `/contributions/:id/comments`
- Create an endpoint for Deleting a comment by id `Delete` `/comments/:comment_id`
- Write tests for the endpoints

### How should this be manually tested?

- At the root of the project directory run the following command:
``` bash
  pytest -v or pytest
```
### What are the relevant pivotal tracker stories?

##3699938

njirap commented:
@andela-oowonikoko my heart is not at peace with the first endpoint `/api/v1/contributions/?include=comments` Is it very necessary? cc @andela-kakpobome

andela-oowonikoko commented:
@njirap According to the story on pivotal-tracker (https://www.pivotaltracker.com/story/show/150002309) by @andela-kakpobome, this endpoint is necessary to get all contributions with their comments.

njirap commented:
![char_limit](https://user-images.githubusercontent.com/2056512/29277636-ddf3fb58-811a-11e7-9068-f3ff3c6576fd.png)


git titles have a character limit.. that's why this title is overflowing into the body and the sentence is not complete. Shorten the title while maintaining it's correctness as a grammatical expression

njirap commented:
get rid of the endpoint ` /api/v1/contributions/?include=comments` reference the slack discussion we had on the same

njirap commented:
![char](https://user-images.githubusercontent.com/2056512/29278071-4c8fa674-811c-11e7-8283-9f4e147ad0ae.png)

just noticed this after merging it to develop. I will be renaming it to shroten it

andela-oowonikoko commented:
The slack discussion says we should create the endpoint `contributions/:id/comments` instead of having to include `comments` into the contributions endpoint. And we should also include `comment_count` as part of the response. I'll work on that now.

andela-oowonikoko commented:
@njirap I have removed the endpoint `/api/v1/contributions/?include=comments` and added `comment_count` to the response.

I have also squashed the commits and renamed it

njirap commented on api/views/contribution_comments.py:
```diff
@@ -0,0 +1,83 @@
+from uuid import uuid4
```
I am moving all this into the contributions views

njirap commented on api/views/contributions.py:
```diff
@@ -34,6 +41,8 @@ def get(self):
         for c in _contributions:
             contrib = c.serialize()
+            contrib['commentCount'] = len(list(map(serialize_thread_comments,
+                                             c.comments)))
```
what's wrong with `len(c.comments)`?

andela-oowonikoko replied:
Fixed. I tried `len(c.comments)` and I was getting an error so I researched further into the error and all I had to do was `len(c.comments.all())`

njirap commented on api/models.py:
```diff
@@ -50,6 +49,23 @@ class Tag(db.Model, SerializerMixin):
+    contribution_id = db.Column('contribution_id', db.String,
+                                     db.ForeignKey('Contribution.id'))
+    message = db.Column(db.String(), nullable=False)
+    parent_id = db.Column(db.String(), db.ForeignKey('Comment.id'))
```
whatt is this field for?

andela-oowonikoko replied:
The `parent_id` field is meant to be `null` for regular comments and meant to have an `id` for replies to a comment

njirap commented on api/models.py:
```diff
@@ -50,6 +49,23 @@ class Tag(db.Model, SerializerMixin):
+
+    def __repr__(self):
+         return '<Comment %r>' % self.message
 
```
2 blank spaces instead of 1

andela-oowonikoko replied:
Fixed 👍

njirap commented on api/models.py:
```diff
@@ -50,6 +49,23 @@ class Tag(db.Model, SerializerMixin):
     def __repr__(self):
         return '<Tag %r>' % self.name
 
```
2 blank spaces instead of 1

andela-oowonikoko replied:
Fixed 👍

njirap commented:
LGTM
